modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/RecnoConnectionManager.java

```diff
@@ -0,0 +1,108 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ * 
+ *    (C) 2002-2015, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.data.shapefile;
+
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.geotools.util.logging.Logging;
+
+/**
+ * Manages a connection pool for the JDBC providers used with the optional RECNO field index.
+ * 
+ * @author Alvaro Huarte - Tracasa / ahuarte@tracasa.es
+ */
+class RecnoConnectionManager
+{
+    static final Logger LOGGER = Logging.getLogger(RecnoConnectionManager.class);
+    
+    // Provides a helper class to share a Connection.
+    private class ConnectionPool
+    {
+        // Creates a new ConnectionPool object.
+        public ConnectionPool(java.sql.Connection connection)
+        {
+            Connection = connection;
+            ReferenceCount = 1;
+        }
+        // The Connection object managed.
+        public java.sql.Connection Connection;
+        // Number of references to the Connection object managed.
+        public int ReferenceCount;
+    }
+    // Hash of ConnectionPools managed.
+    private HashMap<String,ConnectionPool> connectionHash = new HashMap<String,ConnectionPool>();
+    
+    /**
+     * Attempts to establish a connection to the given database URL.
+     */
+    public java.sql.Connection getConnection(String connectionString) throws ClassNotFoundException, SQLException
+    {
+        if (connectionString==null || connectionString.length()==0)
+            return null;
+        
+        synchronized (this) 
+        {
+            ConnectionPool connectionPool = connectionHash.get(connectionString);
+            
+            if (connectionPool==null)
+            {
+                Class.forName("sun.jdbc.odbc.JdbcOdbcDriver");
+                java.sql.Connection connection = java.sql.DriverManager.getConnection(connectionString, "", "");
+                
+                if (connection==null)
+                {
+                    LOGGER.log(Level.SEVERE, "Error getting a shared Connection of '" + connectionString + "'.");
+                    return null;
+                }
+                connectionPool = new ConnectionPool(connection);
+                connectionHash.put(connectionString, connectionPool);
```

Comment:
So you have a single connection used in different threads? This will break under load, the java.sql.Connection is not supposed to be thread safe, as it keeps state. You should use commons-dbpc to manage the connection pool insetad of managing it yourlself, and make sure you clean them up when the store goes down)